# Spam Detector

In [25]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

## Retrieve the Data

The data is located at [https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv](https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv)

Dataset Source: [UCI Machine Learning Library](https://archive.ics.uci.edu/dataset/94/spambase)

Import the data using Pandas. Display the resulting DataFrame to confirm the import was successful.

In [26]:
# Import the data
data = pd.read_csv("https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv")
data.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0            0.00               0.64           0.64           0.0   
1            0.21               0.28           0.50           0.0   
2            0.06               0.00           0.71           0.0   
3            0.00               0.00           0.00           0.0   
4            0.00               0.00           0.00           0.0   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0           0.32            0.00              0.00                0.00   
1           0.14            0.28              0.21                0.07   
2           1.23            0.19              0.19                0.12   
3           0.63            0.00              0.31                0.63   
4           0.63            0.00              0.31                0.63   

   word_freq_order  word_freq_mail  ...  char_freq_;  char_freq_(  \
0             0.00            0.00  ...         0.00        0.000   
1             0.00            0.94  ...         0.00        0.132   
2             0.64            0.25  ...         0.01        0.143   
3             0.31            0.63  ...         0.00        0.137   
4             0.31            0.63  ...         0.00        0.135   

   char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0          0.0        0.778        0.000        0.000   
1          0.0        0.372        0.180        0.048   
2          0.0        0.276        0.184        0.010   
3          0.0        0.137        0.000        0.000   
4          0.0        0.135        0.000        0.000   

   capital_run_length_average  capital_run_length_longest  \
0                       3.756                          61   
1                       5.114                         101   
2                       9.821                         485   
3                       3.537                          40   
4                       3.537                          40   

   capital_run_length_total  spam  
0                       278     1  
1                      1028     1  
2                      2259     1  
3                       191     1  
4                       191     1  

[5 rows x 58 columns]

## Predict Model Performance

You will be creating and comparing two models on this data: a Logistic Regression, and a Random Forests Classifier. Before you create, fit, and score the models, make a prediction as to which model you think will perform better. You do not need to be correct!

Write down your prediction in the designated cells in your Jupyter Notebook, and provide justification for your educated guess.

# Prediction:

I would predict that the Random Forest model will perform better with the data I will be working with above. This model is used for large amounts of data that it looks like I will be working with. It can handle this high-dimensonal data (as it appears), while the Logistic Regression model is sometimes more unpredictable for usage, as it is prone to overfitting. Random Forest models also work by:
>"averaging the predictions of several trees rather than depending just on one, it first lowers overfitting"

Logistic Regression models are limited to modeling their assumptions of linear relationships between the variables. Although the two classifications of "spam" and "not spam" are what will ultimately be the detected outcome after using one model or the other, the Random Forest model can look at the more complex relationships between my variables, for a possibly more accurate result.


Reference for above: [Logistic Regression Vs Random Forest Classifier](https://www.geeksforgeeks.org/logistic-regression-vs-random-forest-classifier/)





## Split the Data into Training and Testing Sets

In [27]:
# A peak at the dataframe (with value_counts())
data.value_counts()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  word_freq_order  word_freq_mail  word_freq_receive  word_freq_will  word_freq_people  word_freq_report  word_freq_addresses  word_freq_free  word_freq_business  word_freq_email  word_freq_you  word_freq_credit  word_freq_your  word_freq_font  word_freq_000  word_freq_money  word_freq_hp  word_freq_hpl  word_freq_george  word_freq_650  word_freq_lab  word_freq_labs  word_freq_telnet  word_freq_857  word_freq_data  word_freq_415  word_freq_85  word_freq_technology  word_freq_1999  word_freq_parts  word_freq_pm  word_freq_direct  word_freq_cs  word_freq_meeting  word_freq_original  word_freq_project  word_freq_re  word_freq_edu  word_freq_table  word_freq_conference  char_freq_;  char_freq_(  char_freq_[  char_freq_!  char_freq_$  char_freq_#  capital_run_length_average  capital_run_length_longest  capital_run_length_total  spam
0.00            0.00         

In [28]:
# Create the labels set `y` and features DataFrame `X`
y = data['spam']
X = data.drop('spam', axis=1)
# Checking that 'y' is only the 'spam' column
print(y)

0       1
1       1
2       1
3       1
4       1
       ..
4596    0
4597    0
4598    0
4599    0
4600    0
Name: spam, Length: 4601, dtype: int64


In [29]:
# Check the balance of the labels variable (`y`) by using the `value_counts` function.
y.value_counts()

spam
0    2788
1    1813
Name: count, dtype: int64

In [30]:
# Split the data into X_train, X_test, y_train, y_test
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
X_train.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
1370            0.09                0.0           0.09           0.0   
3038            0.00                0.0           0.00           0.0   
2361            0.00                0.0           2.43           0.0   
156             0.00                0.0           0.00           0.0   
2526            0.00                0.0           0.00           0.0   

      word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
1370           0.39            0.09              0.09                0.00   
3038           0.00            0.00              0.00                0.00   
2361           0.00            0.00              0.00                0.00   
156            1.31            0.00              1.31                1.31   
2526           0.00            0.00              0.00                0.00   

      word_freq_order  word_freq_mail  ...  word_freq_conference  char_freq_;  \
1370             0.19            0.29  ...                   0.0        0.000   
3038             0.00            0.00  ...                   0.0        0.124   
2361             0.27            0.00  ...                   0.0        0.000   
156              1.31            1.31  ...                   0.0        0.000   
2526             0.00            0.00  ...                   0.0        0.000   

      char_freq_(  char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
1370        0.139          0.0        0.310        0.155          0.0   
3038        0.124          0.0        0.000        0.000          0.0   
2361        0.344          0.0        0.000        0.000          0.0   
156         0.000          0.0        0.117        0.117          0.0   
2526        0.000          0.0        0.000        0.000          0.0   

      capital_run_length_average  capital_run_length_longest  \
1370                       6.813                         494   
3038                       1.800                           8   
2361                       2.319                          12   
156                       48.500                         186   
2526                       2.307                           8   

      capital_run_length_total  
1370                      1458  
3038                        45  
2361                       167  
156                        291  
2526                        30  

[5 rows x 57 columns]

## Scale the Features

Use the `StandardScaler` to scale the features data. Remember that only `X_train` and `X_test` DataFrames should be scaled.

In [31]:
from sklearn.preprocessing import StandardScaler

# Create the StandardScaler instance
scaler = StandardScaler()

In [32]:
# Fit the Standard Scaler with the training data
X_train_scaled = scaler.fit_transform(X_train)

In [33]:
# Scale the training data
X_test_scaled = scaler.transform(X_test)

## Create and Fit a Logistic Regression Model

Create a Logistic Regression model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like.

In [34]:
# Train a Logistic Regression model and print the model score
from sklearn.linear_model import LogisticRegression

lr_model = LogisticRegression(random_state=1)

# Fitting to the scaled training data
lr_model.fit(X_train_scaled, y_train)

LogisticRegression(random_state=1)

In [35]:
# Make and save testing predictions with the saved logistic regression model using the test data
testing_predictions = lr_model.predict(X_test_scaled)
# Review the predictions
print(testing_predictions)

[0 0 0 1 0 1 0 0 0 1 0 0 0 1 0 0 0 0 1 0 0 1 0 0 1 0 1 0 1 1 1 1 0 1 0 0 1
 1 0 1 1 0 0 0 0 1 1 0 0 1 1 0 0 0 0 1 0 0 0 0 1 0 1 1 1 0 0 0 0 0 0 1 0 1
 1 0 0 0 0 1 0 0 0 0 0 0 1 1 1 0 1 1 0 0 0 0 1 0 1 1 0 1 0 1 1 0 0 1 1 1 1
 1 1 0 1 1 0 0 0 0 1 0 1 0 1 0 0 0 1 0 0 1 1 0 0 0 1 0 1 1 1 1 0 0 1 1 0 0
 0 0 0 1 1 1 0 0 0 1 1 0 1 1 1 0 1 0 1 0 0 1 0 0 0 0 1 0 1 0 0 1 0 0 0 0 1
 1 0 0 0 0 1 0 1 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 1 0 1 0 0 0 0 0 0 0 0 0 1
 1 1 1 0 1 0 0 0 0 0 0 0 0 0 0 0 1 1 0 1 1 1 0 0 0 1 0 0 0 1 0 0 0 0 0 1 1
 0 1 0 1 0 1 1 1 0 0 0 1 0 0 0 0 0 1 0 0 0 0 1 0 0 1 0 0 1 0 1 0 1 0 0 1 1
 0 0 0 0 0 1 0 0 0 0 1 1 1 0 1 0 1 1 0 0 0 1 1 0 0 0 0 0 0 1 1 1 1 0 0 0 1
 0 0 1 0 1 0 0 0 0 0 0 1 0 0 0 1 1 0 1 0 0 0 1 0 1 1 1 0 1 1 0 0 0 0 0 0 0
 1 1 0 1 0 0 0 0 1 0 1 1 0 0 1 0 1 0 0 0 0 0 1 1 0 1 0 1 0 0 1 0 0 0 1 1 0
 1 1 1 0 0 1 0 0 1 0 0 0 1 0 1 0 0 0 0 1 1 0 0 0 1 0 0 1 1 0 0 1 0 0 0 1 1
 0 1 0 0 0 1 0 0 0 1 0 0 1 1 1 1 1 1 0 0 1 0 1 1 1 1 0 0 0 0 0 0 0 0 0 0 1
 0 1 1 1 0 0 0 0 0 1 1 1 

In [36]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.

# Referencing "Model and Fit to a Logistic Regression Classifier" in Mod.13, Day 2, Activity 6

acc_score = accuracy_score(y_test, testing_predictions)
print(f"Accuracy Score : {acc_score}")

Accuracy Score : 0.9196525515743756


In [37]:
# An alternative way to "Validate the model by checking the model accuracy with model.score"

# Referencing "Model and Fit to a Logistic Regression Classifier" in Mod.13, Day 3, Activity 3

print('Training Data Score: %.3f' % lr_model.score(X_train_scaled, y_train))
print('Testing Data Score: %.3f' % lr_model.score(X_test_scaled, y_test))

Training Data Score: 0.927
Testing Data Score: 0.920


## Create and Fit a Random Forest Classifier Model

Create a Random Forest Classifier model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like.

In [38]:
# Train a Random Forest Classifier model and print the model score
# Referencing "Fitting the Random Forest Model" in Mod.13, Day 3, Activity 3
from sklearn.ensemble import RandomForestClassifier
rf_model = RandomForestClassifier(n_estimators=128, random_state=1) # I used these settings based on the activity I referenced.

# Fit the model
rf_model.fit(X_train_scaled, y_train)

RandomForestClassifier(n_estimators=128, random_state=1)

In [39]:
# Make and save testing predictions with the saved logistic regression model using the test data
testing_predictions = rf_model.predict(X_test_scaled)

# Review the predictions
testing_predictions

array([0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 1, 1, 0, 0, 1,
       1, 0, 1, 0, 1, 0, 1, 1, 1, 1, 0, 0, 0, 0, 1, 1, 0, 1, 1, 0, 0, 0,
       1, 1, 1, 0, 0, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 1, 0,
       0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1,
       1, 0, 1, 1, 0, 1, 0, 0, 1, 0, 1, 1, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0,
       0, 1, 1, 0, 1, 1, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 0, 1,
       1, 0, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 0, 0, 0, 1, 0, 1, 1, 0,
       0, 0, 0, 1, 1, 0, 1, 1, 1, 0, 1, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0,
       1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1,
       1, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 1, 0,
       1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 0, 1, 0,
       0, 1, 0, 1, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0,

In [40]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
acc_score = accuracy_score(y_test, testing_predictions)
print(f"Accuracy Score : {acc_score}")

Accuracy Score : 0.9554831704668838


## Evaluate the Models

Which model performed better? How does that compare to your prediction? Write down your results and thoughts in the following markdown cell.

### Answer:
1. The Random Forest model performed better. It had an accuracy score of **0.9554831704668838 (about 96%)**, when, compared to the Logistic Regression score of **Accuracy Score : 0.9196525515743756 (about 92%)**, it is a significant improvement. It shows that there are likely complex patterns or interactions between features in the spam dataset.
2. It looks like my prediction of "Random Forest model" was accurate, although I had no precise way of knowing how these two models would perform with this specific dataset, at the first glance. Like I stated in my prediction, the assumption of a linear relationship between features and the target variable, that the Logistic Regression model makes, may play a large role in this models ability to classify spam. This would potentially cause this model to struggle when it finds data that does not fit the linear framework.

Reference: The above reference in the "Prediction" text box, as well as [Machine Learning Techniques for Spam Detection in Email](https://medium.com/@alinatabish/machine-learning-techniques-for-spam-detection-in-email-7db87eb11bc2) - Alina Tabish